File: php/nodes/resolve_urls.php
<?php

require_once __DIR__ . '/../utils/resolve_khinsider_url.php';

function resolve_urls(array $edges, array $utils = []): callable {
    $utils += ['resolver' => 'resolve_khinsider_url'];

    return function ($state) use ($edges, $utils): Next {
        echo "Resolving Khinsider URLs...\n";

        $resolved = array_map(function ($url) use ($utils) {
            if (str_starts_with($url, 'https://downloads.khinsider.com/game-soundtracks')) {
                return ($utils['resolver'])($url);
            }
            return $url;
        }, $state['urls']);

        echo "Finished resolving URLs.\n";
        return new Next($edges['on_success'], array_merge($state, ['urls' => array_values($resolved)]));
    };
}

File: ruby/nodes/resolve_urls.rb
# frozen_string_literal: true

require 'async/barrier'
require_relative '../lib/ambler'
require_relative '../utils/resolve_khinsider_url'

module ResolveUrls
  DEFAULT_UTILS = { resolver: ResolveKhinsiderUrl.method(:resolve) }.freeze

  def self.node(edges, utils = DEFAULT_UTILS)
    ->(state) {
      puts 'Resolving Khinsider URLs...'

      resolved_urls = Array.new(state[:urls].length)
      barrier = Async::Barrier.new

      state[:urls].each_with_index do |url, i|
        barrier.async do
          resolved_urls[i] = if url.start_with?('https://downloads.khinsider.com/game-soundtracks')
            utils[:resolver].call(url)
          else
            url
          end
        end
      end

      barrier.wait
      puts 'Finished resolving URLs.'
      Ambler::Next.new(edges[:on_success], state.merge(urls: resolved_urls))
    }
  end
end
